lib/kvs/kvs_cursor.c

```diff
@@ -1100,32 +1100,63 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
     return err;
 }
 
-static merr_t
-ikvs_cursor_kv_copy(struct kvs_cursor_impl *cursor, struct kvs_kvtuple *kvt)
+void
+kvs_cursor_key_copy(struct hse_kvs_cursor *cursor, const void *kbuf, size_t kbufsz, const void **key_out, size_t *klen_out)
 {
-    struct kvs_vtuple *vt = &cursor->kci_elem_last.kce_vt;
-    uint               clen = cursor->kci_elem_last.kce_complen;
-    void *             vbuf = cursor->kci_buf + HSE_KVS_KEY_LEN_MAX;
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+    void                   *buf = (void *)kbuf;
+    size_t                  bufsz = kbufsz;
+    uint                    keylen;
+    const void             *key;
+
+    if (!buf) {
+        buf = (void *)cur->kci_buf;
+        bufsz = HSE_KVS_KEY_LEN_MAX;
+    }
+
+    key = key_obj_copy(buf, bufsz, &keylen, cur->kci_last);
+    *klen_out = keylen;
+
+    if (key_out)
+        *key_out = key;
+}
+
+merr_t
+kvs_cursor_val_copy(struct hse_kvs_cursor *cursor, const void *vbuf, size_t vbufsz, const void **val_out, size_t *vlen_out)
+{
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+
+    struct kvs_vtuple *vt = &cur->kci_elem_last.kce_vt;
+    uint               clen = cur->kci_elem_last.kce_complen;
+    void              *buf = (void *)vbuf;
+    size_t             bufsz = vbufsz;
     merr_t             err = 0;
 
-    kvt->kvt_key.kt_data = key_obj_copy(
-        cursor->kci_buf, HSE_KVS_KEY_LEN_MAX, (uint *)&kvt->kvt_key.kt_len, cursor->kci_last);
+    if (!buf) {
+        buf = cur->kci_buf + HSE_KVS_KEY_LEN_MAX;
+        bufsz = HSE_KVS_VALUE_LEN_MAX;
+    }
+
+    if (vlen_out)
+        *vlen_out = vt->vt_xlen;
 
     if (clen) {
         uint outlen;
 
-        err = compress_lz4_ops.cop_decompress(vt->vt_data, clen, vbuf, HSE_KVS_VALUE_LEN_MAX, &outlen);
+        err = compress_lz4_ops.cop_decompress(vt->vt_data, clen, buf, bufsz, &outlen);
         if (ev(err))
             return err;
 
-        if (ev(outlen != vt->vt_xlen))
-            return merr(EBUG);
+        if (ev(outlen != min_t(u64, vt->vt_xlen, bufsz)))
+            return EBUG;
     } else {
-        memcpy(vbuf, vt->vt_data, vt->vt_xlen);
+        memcpy(buf, vt->vt_data, min_t(u64, vt->vt_xlen, bufsz));
```

Comment:
Use uint64_t

Reply:
This function (kvs_cursor_val_copy) seems like it misuses `vt_xlen`.  When a value is compressed, doesn't `vt_xlen == (clen<<32)|vlen`?

Reply:
The reason this works is because vtuples that come out of cursor merge loops use vt_xlen differently.  Not going to address that as part of this PR.